### PR TITLE
Drag and drop

### PR DESCRIPTION
**Pull request by czyzu0507**

### CODE REVIEW

**Comment by YogAzathoth** [comment]
Closed as for conflicts with master.
New PR will be created when conflicts will be resolved.

**Comment by YogAzathoth** [comment]
Reopenned, no conflicts

File: GUI/src/main/java/Gui.java [hunk 1]
```diff
@@ -7,22 +7,23 @@
...
-        primaryStage.setTitle("Battleship");
-        primaryStage.setScene(scene);
-        primaryStage.show();}
-        catch(Exception e){}
+            Scene scene = new Scene(root, 1200, 800);
+            primaryStage.setResizable(false);
+            primaryStage.setTitle("Battleship");
+            primaryStage.setScene(scene);
+            primaryStage.show();
+        } catch (Exception e) {
```
Comment (clinton56) [inline GUI/src/main/java/Gui.java]: You should not catch general `Exception`. You need to be sure what kind of exception you expect, here it is `IOException`.
You need to catch it and react somehow (error log, closing app?).

File: GUI/src/main/java/components/Ship.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            fxmlLoader.load();
+        } catch (IOException exception) {
+            throw new RuntimeException(exception);
+        }
+    }
+
+
+    @Override
+    public Ship clone() {
+        return new Ship();
```
Comment (clinton56) [inline GUI/src/main/java/components/Ship.java]: This is not good. If calling constructor would be enough then you don't need `clone()`. [http://www.javatpoint.com/object-cloning](url)

File: GUI/src/main/java/controller/GuiController.java [hunk 3]
```diff
@@ -1,44 +1,128 @@
...
+    private DataFormat dataFormat = new DataFormat("Ship");
+    private EventHandler<DragEvent> shipDragOverRoot;
+    private EventHandler<DragEvent> shipDragDropped;
+    private EventHandler<DragEvent> shipDragOverUserBoard;
 
+    @FXML
+    private void fireAway(ActionEvent event) {
+        Node source = (Node) event.getSource();
+        source.setOpacity(1.0);
+        System.out.println("Hello " + "This is the row : " + GridPane.getRowIndex(source) + "\nAnd column: " + GridPane.getColumnIndex(source));
```
Comment (clinton56) [inline GUI/src/main/java/controller/GuiController.java]: You should use logging.

File: GUI/src/main/java/controller/GuiController.java [hunk 4]
```diff
@@ -1,44 +1,128 @@
...
     @FXML
-    private void fireAway(ActionEvent event){
-        Node source = (Node)event.getSource();
+    public void initialize() {
+        userBoard.toFront();
+        Ship ship = new Ship();
+        shipsContainer.getChildren().add(ship);
+        addDragDetection(ship);
+        buildDragHandlers();
+        for (int i = 0; i < 10; i++) {
```
Comment (clinton56) [inline GUI/src/main/java/controller/GuiController.java]: :8ball:  Magical number `10` should be extracted some constant.

File: GUI/src/main/java/controller/GuiController.java [hunk 5]
```diff
@@ -1,44 +1,128 @@
...
+            for (int j = 0; j < 10; j++) {
+                Pane label = new Pane();
+                label.setOnDragDropped(event -> {
+                    event.acceptTransferModes(TransferMode.ANY);
+                    System.out.println(GridPane.getRowIndex(label) + " " +GridPane.getColumnIndex(label));
```
Comment (clinton56) [inline GUI/src/main/java/controller/GuiController.java]: Logger wanted :dancers:

File: GUI/src/main/java/controller/GuiController.java [hunk 6]
```diff
@@ -1,44 +1,128 @@
...
-        System.out.println("Hello " + "This is the row : "+GridPane.getRowIndex(source)+"\nAnd column: "+ GridPane.getColumnIndex(source));
+    private void addDragDetection(Ship ship) {
 
+        ship.setOnDragDetected(new EventHandler<MouseEvent>() {
 
-        source.setDisable(true);
-        }
+            @Override
+            public void handle(MouseEvent event) {
+                System.out.println("test");
```
Comment (clinton56) [inline GUI/src/main/java/controller/GuiController.java]: So many responsibilities in one method and also too many lines of code. Uncle Bob wouldn't be happy

File: GUI/src/main/java/controller/GuiController.java [hunk 7]
```diff
@@ -1,44 +1,128 @@
...
+            public void handle(DragEvent event) {
+                event.acceptTransferModes(TransferMode.ANY);
+                invship.relocate(event.getSceneX() - invship.getWidth() / 2, event.getSceneY() - invship.getHeight() / 4);
+                event.consume();
+            }
+        };
+
+        shipDragOverUserBoard = new EventHandler<DragEvent>() {
+
+            @Override
```
Comment (clinton56) [inline GUI/src/main/java/controller/GuiController.java]: DRY : lines 99-105 and 109-115 does the same. You could create private method and pass there `Event object`.

File: GUI/src/main/resources/gui.fxml [hunk 8]
```diff
@@ -6,209 +6,198 @@
...
+        <Pane layoutX="100.0" layoutY="50.0" prefHeight="420.0" prefWidth="420.0">
+            <children>
+                <ImageView pickOnBounds="true" preserveRatio="true">
+                    <image>
+                        <Image url="@assets/Grid.png" />
+                    </image>
+                </ImageView>
+                <GridPane fx:id="userBoard" layoutX="10.0" layoutY="10.0" prefHeight="400.0" prefWidth="400.0">
+                    <columnConstraints>
+                        <ColumnConstraints hgrow="SOMETIMES" maxWidth="40.0" minWidth="40.0" prefWidth="40.0" />
```
Comment (clinton56) [inline GUI/src/main/resources/gui.fxml]: :ghost: Why is it repeated so many times?

File: GUI/src/main/resources/gui.fxml [hunk 9]
```diff
@@ -6,209 +6,198 @@
...
-               </image>
-            </ImageView>
-         </children>
-      </Pane>
-   </children>
-</Pane>
+                  <RowConstraints />
+                    </rowConstraints>
+                    <children>
+                        <Button mnemonicParsing="false" onAction="#fireAway" opacity="0.0" prefHeight="40.0" prefWidth="40.0" styleClass="buttons" stylesheets="@buttons.css" textAlignment="CENTER" textFill="WHITE" GridPane.columnIndex="1" GridPane.rowIndex="1" />
```
Comment (clinton56) [inline GUI/src/main/resources/gui.fxml]: What? Why so many items?

Comment (czyzu0507) [inline GUI/src/main/resources/gui.fxml]: It is button for every cell in grid pane.

File: GUI/src/main/resources/gui.fxml (hunk 8 above)
Comment (czyzu0507) [inline GUI/src/main/resources/gui.fxml]: scene builder did it like this.

File: GUI/src/main/java/components/Ship.java (hunk 2 above)
Comment (czyzu0507) [inline GUI/src/main/java/components/Ship.java]: I created it for the future

File: GUI/src/main/java/Gui.java (hunk 1 above)
Comment (czyzu0507) [inline GUI/src/main/java/Gui.java]: I already changed it.

File: GUI/src/main/java/controller/GuiController.java (hunk 4 above)
Comment (czyzu0507) [inline GUI/src/main/java/controller/GuiController.java]: I created final variable BORD_SIZE

Comment (LIttleAncientForestKami) [inline GUI/src/main/java/controller/GuiController.java]: Hopefully B_OA_RD rather than BORD? :-)

File: GUI/src/main/java/Gui.java [hunk 10]
```diff
@@ -4,25 +4,29 @@
...
-        primaryStage.setResizable(false);
-        primaryStage.setTitle("Battleship");
-        primaryStage.setScene(scene);
-        primaryStage.show();}
-        catch(Exception e){}
-    }
-
-    public static void main(String[] args){
-        launch(args);
+            Scene scene = new Scene(root, 1200, 800);
```
Comment (LIttleAncientForestKami) [inline GUI/src/main/java/Gui.java]: magic :1234:

File: GUI/src/main/java/Gui.java [hunk 11]
```diff
@@ -4,25 +4,29 @@
...
+            primaryStage.setResizable(false);
+            primaryStage.setTitle("Battleship");
+            primaryStage.setScene(scene);
+            primaryStage.show();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (LIttleAncientForestKami) [inline GUI/src/main/java/Gui.java]: Error handling was forgotten?

**Comment by LIttleAncientForestKami** [comment]
How can I verify this works? How can I test this? What this is supposed to do? When can I say, yeah, this really does what it should, let's merge?

**Comment by czyzu0507** [comment]
You wont me to add description for this or what

**Pull request closed without merging**